apps/src/templates/instructions/ReviewTab.jsx

```diff
@@ -68,56 +70,86 @@ class ReviewTab extends Component {
+            this.setState({
+              comments: data,
+              token: request.getResponseHeader('csrf-token')
+            });
+            resolve();
+          });
+      })
+    );
 
-    codeReviewDataApi
-      .getPeerReviewStatus(channelId, serverLevelId, serverScriptId)
-      .done(data => {
-        const id = (data && data.id) || null;
-        this.setState({
-          reviewCheckboxEnabled: data.canMarkReviewable,
-          isReadyForReview: data.reviewEnabled,
-          projectOwnerName: data.name,
-          reviewableProjectId: id
-        });
+    initialLoadPromises.push(
+      new Promise((resolve, reject) => {
+        codeReviewDataApi
+          .getPeerReviewStatus(channelId, serverLevelId, serverScriptId)
+          .done(data => {
+            const id = (data && data.id) || null;
+            this.setState({
+              reviewCheckboxEnabled: data.canMarkReviewable,
+              isReadyForReview: data.reviewEnabled,
+              projectOwnerName: data.name,
+              reviewableProjectId: id
+            });
+            resolve();
+          })
+          .fail(() => {
+            this.setState({
+              reviewCheckboxEnabled: false,
+              isReadyForReview: false
+            });
+            reject();
+          });
       })
-      .fail(() => {
-        this.setState({
-          reviewCheckboxEnabled: false,
-          isReadyForReview: false
-        });
-      });
+    );
 
     if (
       !this.props.viewAsCodeReviewer &&
       this.props.viewAs !== ViewType.Teacher
     ) {
-      codeReviewDataApi
-        .getReviewablePeers(channelId, serverLevelId, serverScriptId)
-        .done(data => {
-          this.setState({
-            reviewablePeers: _.chain(data)
-              .filter(peerEntry => peerEntry && peerEntry.length === 2)
-              .map(peerEntry => ({id: peerEntry[0], name: peerEntry[1]}))
-              .value()
-          });
+      initialLoadPromises.push(
+        new Promise((resolve, reject) => {
+          codeReviewDataApi
+            .getReviewablePeers(channelId, serverLevelId, serverScriptId)
+            .done(data => {
+              this.setState({
+                reviewablePeers: _.chain(data)
+                  .filter(peerEntry => peerEntry && peerEntry.length === 2)
+                  .map(peerEntry => ({id: peerEntry[0], name: peerEntry[1]}))
+                  .value()
+              });
+              resolve();
+            })
+            .fail(() => {
+              this.setState({
+                errorLoadingReviewblePeers: true
+              });
+              reject();
+            });
         })
-        .fail(() => {
-          this.setState({
-            errorLoadingReviewblePeers: true
-          });
-        });
+      );
     }
+
+    Promise.all(initialLoadPromises).finally(() => {
+      this.setState({
+        initialLoadCompleted: true
+      });
+      this.props.onLoadComplete();
```

Comment:
This seems like it works based on your video, but is there any chance that we'd end up resizing before the comments have loaded? Or guaranteed to resize after comments have loaded? My async Javascript is still a little hazy...

Reply:
So this code should only execute once all the async calls have completed (either completed or errored). However, looking at this again, it may be better to move the `this.props.onLoadComplete()` to the `componentDidUpdate` hook to guarantee that we fire the resize callback after the state changes have been applied and the component has re-rendered.

Reply:
Yeah that `initialLoadCompleted` state change was the one I wasn't sure about. That sounds like a good update to me.